File: Modularity-State-Management-Architecture/Library.Infrastructure/Library/Infrastructure/IWorkItem.cs
using Newtonsoft.Json;
using StateBuilder.Events;
using System;
using Unity;

namespace StateBuilder.Library.Infrastructure
{
    [JsonObject(ItemReferenceLoopHandling = ReferenceLoopHandling.Serialize, IsReference = true)]
    public interface IWorkItem : IDisposable
    {
        [JsonIgnore]
        Guid Id { get; set; }
        //[JsonProperty(ReferenceLoopHandling = ReferenceLoopHandling.Ignore, IsReference = true)]
        [JsonIgnore]
        IWorkItem Parent { get; }
        //[JsonIgnore]
        //[JsonProperty(ReferenceLoopHandling = ReferenceLoopHandling.Ignore, IsReference = true)]
        [JsonIgnore]
        IWorkItem RootWorkItem { get; }
        [JsonIgnore]
        IUnityContainer Container { get; }
        [JsonProperty]
        State State { get; }
        [JsonProperty]
        WorkItemStatus Status { get; }

        [JsonIgnore]
        WorkItemCollection WorkItems { get; }

        event EventHandler Disposed;
        event EventHandler<StateChangedEventArgs> StateChanged;

        SubscriptionToken SubscribeStateChangedEvent(string stateName, Action action);
        SubscriptionToken SubscribeStateChangedEvent(string stateName, Action<object> action);
        SubscriptionToken SubscribeStateChangedEvent(string stateName, Action<StateChangedEventArgs> action);
        SubscriptionToken SubscribeStateChangedEvent<T>(string stateName, Action<T> action);
        SubscriptionToken SubscribeStateChangedEvent<T>(string stateName, Action<StateChangedEventArgs<T>> action);

        void Activate();
        void Run(Action continueWith = null);
        Type GetTypeOfState(string stateName);

        event EventHandler<BroadCastStateEventArgs> BroadCastEvent;
        void BroadCastState(object sender, BroadCastStateEventArgs eventArgs);
    }
    public static class WorkItemExtensions
    {
        private class __WorkItem : WorkItem
        {
            public __WorkItem(IUnityContainer container)
                : base(container)
            {
            }
        }
        private static IWorkItem _rootWorkItem = null;
        public static IWorkItem RootWorkItem
        {
            get
            {
                if (_rootWorkItem == null)
                {
                    UnityContainer container = new UnityContainer();
                    _rootWorkItem = new __WorkItem(container);
                }
                return _rootWorkItem;
            }
        }
    }
}

File: Modularity-State-Management-Architecture/WebApplicationNG/ClientApp/src/StateBuilders/Library/Infrastructure/IWorkItem.ts
import { EventEmitter } from '@angular/core'; 
import { SubscriptionToken } from '../../Events/SubscriptionToken';
import { v4 as uuid } from 'uuid';
import { State } from './State';
import { StateChangedEventArgs } from './StateChangedEventArgs';
import { WorkItemStatus } from './WorkItemStatus';
export interface IWorkItem{
	Id:uuid;
	Parent:IWorkItem;
	RootWorkItem:IWorkItem;
	//Container:IUnityContainer;
	WIState:State;
	Status:WorkItemStatus;
	WorkItemStateChanged : EventEmitter<StateChangedEventArgs>; 
	SubscribeStateChangedEvent(stateName:string, action:(_currentObject:any)=>void,currentObj:any):SubscriptionToken;
	SubscribeStateChangedEventWithObject(stateName:string, action:(_object:object,_currentObject:any)=>void,currentObj:any):SubscriptionToken;
	SubscribeStateChangedEventWithArgs(stateName:string, action:(_object:StateChangedEventArgs,_currentObject:any)=>any,currentObj:any):SubscriptionToken;
	//Activate(): void;
	//Run(continueWith : ()=>null):void;
	Run():void;
}